File: streams/src/Stream.js
/**
 * Stream — fluent builder for streaming pipelines.
 *
 * A Stream is an immutable chain of Operators that gets compiled and run
 * by a Runner. Each combinator (`.map`, `.filter`, ...) returns a NEW Stream
 * with the operator appended; the chain is finalized by `.run({ queryId })`.
 *
 * The chain shape (operator types + their stateful config) is fingerprinted
 * via configHash so that re-deploying with a different chain under the same
 * queryId is rejected at registration unless `{ reset: true }` is passed.
 */

import { MapOperator } from './operators/MapOperator.js'
import { FilterOperator } from './operators/FilterOperator.js'
import { FlatMapOperator } from './operators/FlatMapOperator.js'
import { KeyByOperator } from './operators/KeyByOperator.js'
import { WindowTumblingOperator } from './operators/WindowTumblingOperator.js'
import { WindowSlidingOperator } from './operators/WindowSlidingOperator.js'
import { WindowSessionOperator } from './operators/WindowSessionOperator.js'
import { WindowCronOperator } from './operators/WindowCronOperator.js'
import { ReduceOperator } from './operators/ReduceOperator.js'
import { AggregateOperator } from './operators/AggregateOperator.js'
import { SinkOperator } from './operators/SinkOperator.js'
import { ForeachOperator } from './operators/ForeachOperator.js'
import { Runner } from './runtime/Runner.js'
import { configHashOf } from './util/configHash.js'

export class Stream {
  /**
   * Build a Stream sourced from a Queen QueueBuilder.
   *
   * @param {object} queueBuilder - the result of queen.queue('name') — must
   *   expose .pop() / .ack() / .push() and the queue name.
   * @param {object} [options]
   * @returns {Stream}
   */
  static from(queueBuilder, options = {}) {
    return new Stream({
      source: queueBuilder,
      operators: [],
      sourceOptions: options
    })
  }

  /**
   * @param {object} args
   * @param {object} args.source - source QueueBuilder
   * @param {object} [args.sourceOptions]
   * @param {Array} args.operators - operator chain
   */
  constructor({ source, operators, sourceOptions = {} }) {
    this._source = source
    this._sourceOptions = sourceOptions
    this._operators = operators
  }

  // ---------------------------------------------------------------- Stateless

  /** @param {(msg:object)=>any|Promise<any>} fn */
  map(fn) {
    return this._extend(new MapOperator(fn))
  }

  /** @param {(msg:object)=>boolean|Promise<boolean>} predicate */
  filter(predicate) {
    return this._extend(new FilterOperator(predicate))
  }

  /** @param {(msg:object)=>(any[]|Promise<any[]>)} fn */
  flatMap(fn) {
    return this._extend(new FlatMapOperator(fn))
  }

  // -------------------------------------------------------------------- Keying

  /**
   * Override the implicit partition-key for downstream stateful operators.
   * If omitted, stateful operators use the source partition_id as the key,
   * which is the natural fit for queues partitioned by entity.
   *
   * Warning: when the key differs from the partition, multiple workers
   * holding different partition leases may write the same logical key,
   * causing cross-worker contention on (query_id, partition_id, key) state
   * rows. The repartition pattern (push to a co-keyed intermediate queue,
   * then process) avoids this.
   *
   * @param {(msg:object)=>string} fn
   */
  keyBy(fn) {
    return this._extend(new KeyByOperator(fn))
  }

  // -------------------------------------------------------------------- Windows

  /**
   * Fixed-size, non-overlapping windows.
   * @param {{
   *   seconds: number,
   *   gracePeriod?: number,    // accept events for windowEnd + grace before closing
   *   idleFlushMs?: number,    // close ripe windows on quiet partitions every N ms (default 5000)
   *   eventTime?: (msg:any)=>(number|Date|string),  // event-time mode if set
   *   allowedLateness?: number, // event-time only: drop events older than wm - allowedLateness
   *   onLate?: 'drop' | 'include'
   * }} opts
   */
  windowTumbling(opts) {
    return this._extend(new WindowTumblingOperator(opts))
  }

  /**
   * Overlapping fixed-size windows that hop every `slide` seconds.
   * Each event creates `size/slide` state rows per key.
   * @param {{size:number, slide:number, gracePeriod?:number, idleFlushMs?:number,
   *          eventTime?:(msg:any)=>any, allowedLateness?:number, onLate?:string}} opts
   */
  windowSliding(opts) {
    return this._extend(new WindowSlidingOperator(opts))
  }

  /**
   * Per-key activity-based windows. A session for a key extends as long
   * as events keep arriving within `gap` seconds of each other.
   * @param {{gap:number, gracePeriod?:number, idleFlushMs?:number,
   *          eventTime?:(msg:any)=>any, allowedLateness?:number, onLate?:string}} opts
   */
  windowSession(opts) {
    return this._extend(new WindowSessionOperator(opts))
  }

  /**
   * Wall-clock-aligned windows. Currently supports the `every:` shorthand
   * (`'second'|'minute'|'hour'|'day'|'week'`); full cron syntax is reserved
   * for v0.3.
   * @param {{every:string, gracePeriod?:number, idleFlushMs?:number,
   *          eventTime?:(msg:any)=>any, allowedLateness?:number, onLate?:string}} opts
   */
  windowCron(opts) {
    return this._extend(new WindowCronOperator(opts))
  }

  // -------------------------------------------------------------------- Reduce

  /**
   * @template T
   * @param {(acc:T, msg:object)=>T|Promise<T>} fn
   * @param {T} initial
   */
  reduce(fn, initial) {
    return this._extend(new ReduceOperator(fn, initial))
  }

  /**
   * Sugar over reduce. Each provided extractor produces one named field on
   * the aggregate value.
   *
   * Example:
   *   .aggregate({ count: () => 1, sum: m => m.data.amount })
   *
   * Output value shape: { count: number, sum: number, ... }
   *
   * @param {Record<string,(msg:object)=>number>} extractors
   */
  aggregate(extractors) {
    return this._extend(new AggregateOperator(extractors))
  }

  // ---------------------------------------------------------------------- Sink

  /**
   * Sink to a Queen queue. The cycle commits state + push + ack atomically.
   *
   * @param {object} sinkQueueBuilder - `queen.queue('sink-name')`
   * @param {{partition?: string|((value:any)=>string)}} [opts]
   */
  to(sinkQueueBuilder, opts = {}) {
    return this._extend(new SinkOperator(sinkQueueBuilder, opts))
  }

  /**
   * Terminal at-least-once side-effect. The cycle acks the source only after
   * `fn` has resolved successfully, so a crash mid-`fn` will redeliver. If
   * you need exactly-once external effects, write to a sink queue and have
   * a dedicated worker consume it.
   *
   * @param {(value:any)=>void|Promise<void>} fn
   */
  foreach(fn) {
    return this._extend(new ForeachOperator(fn))
  }

  // ----------------------------------------------------------------- Compile/run

  /**
   * Start the streaming runner. Returns a handle exposing:
   *   - stop(): graceful drain
   *   - metrics(): cycle/throughput/lag stats
   *
   * @param {object} runOptions
   * @param {string} runOptions.queryId - durable identity of this query
   * @param {number} [runOptions.batchSize=200] - messages per cycle
   * @param {number} [runOptions.maxPartitions=4] - lease up to N partitions/cycle
   * @param {number} [runOptions.maxWaitMillis=1000] - long-poll wait for source pop
   * @param {string} [runOptions.subscriptionMode] - 'all' (default) | 'new'
   * @param {string} [runOptions.subscriptionFrom] - ISO timestamp or 'now'
   * @param {boolean} [runOptions.reset=false] - wipe state on config_hash mismatch
   * @param {(err:Error, ctx:object)=>void} [runOptions.onError] - cycle error hook
   * @param {AbortSignal} [runOptions.abortSignal] - external cancellation
   */
  async run(runOptions) {
    if (!runOptions || !runOptions.queryId) {
      throw new Error('run({ queryId }) is required')
    }
    const compiled = this._compile()
    const runner = new Runner({
      ...runOptions,
      stream: compiled
    })
    await runner.start()
    return runner
  }

  // ===== internals =========================================================

  /** @returns {Stream} */
  _extend(operator) {
    return new Stream({
      source: this._source,
      sourceOptions: this._sourceOptions,
      operators: [...this._operators, operator]
    })
  }

  /**
   * Compile the operator chain into a runnable description: stages, sink,
   * and the configHash. Exposed for the Runner.
   */
  _compile() {
    const sink = this._operators.find(op => op.kind === 'sink' || op.kind === 'foreach')
    const sinkIdx = sink ? this._operators.indexOf(sink) : this._operators.length

    // Verify the sink (if present) is the last operator.
    if (sink && sinkIdx !== this._operators.length - 1) {
      throw new Error(
        'sink operators (.to / .foreach) must be the last in the chain; ' +
        `found ${sink.kind} at position ${sinkIdx} of ${this._operators.length}`
      )
    }

    const upstream = sink ? this._operators.slice(0, sinkIdx) : this._operators

    // Decompose into stages:
    //   pre:    [stateless...] applied to each source record
    //   keyBy:  optional KeyByOperator (operates on source record)
    //   window: optional WindowTumblingOperator (annotates envelopes)
    //   reducer: optional ReduceOperator/AggregateOperator
    //   post:   [stateless...] applied to each emit value (post-reducer)
    //   sink:   optional SinkOperator/ForeachOperator (terminal)
    //
    // "phase" walks from `pre` -> `keyed` -> `window` -> `reducer` -> `post`.
    // Stateless operators belong to whichever side of the reducer they're
    // declared on. A reducer is required to switch to `post`; without one,
    // all stateless ops are pre-stage.
    const stages = {
      pre: [],
      keyBy: null,
      window: null,
      reducer: null,
      post: [],
      sink
    }
    let phase = 'pre'
    for (const op of upstream) {
      if (op.kind === 'map' || op.kind === 'filter' || op.kind === 'flatMap') {
        if (phase === 'pre' || phase === 'keyed') {
          stages.pre.push(op)
        } else if (phase === 'window') {
          // Stateless ops between window and reducer are unusual; treat as pre
          // (they see the windowed envelope but have no reducer state to act on).
          stages.pre.push(op)
        } else {
          // post-reducer: operates on each emit value.
          stages.post.push(op)
        }
      } else if (op.kind === 'keyBy') {
        if (stages.keyBy) throw new Error('only one .keyBy() per stream')
        if (phase === 'reducer') throw new Error('.keyBy() must come before window/reduce')
        stages.keyBy = op
        if (phase === 'pre') phase = 'keyed'
      } else if (op.kind === 'window') {
        if (stages.window) throw new Error('only one window operator per stream')
        if (phase === 'reducer') throw new Error('window must come before reduce/aggregate')
        stages.window = op
        phase = 'window'
      } else if (op.kind === 'reduce' || op.kind === 'aggregate') {
        if (stages.reducer) throw new Error('only one reduce/aggregate per stream')
        if (!stages.window) {
          throw new Error('reduce/aggregate requires a preceding window operator')
        }
        stages.reducer = op
        phase = 'reducer'
      } else {
        throw new Error(`unsupported operator: ${op.kind}`)
      }
    }
    // Backwards-compat: keep the old name `stateless` aliasing `pre` so any
    // existing code paths that referenced stages.stateless keep working.
    stages.stateless = stages.pre

    // Compute config_hash from the chain shape (operator kinds + their
    // structural config; user functions are intentionally not hashed since
    // we cannot stably serialise closures).
    const config_hash = configHashOf(this._operators)

    return {
      source: this._source,
      sourceOptions: this._sourceOptions,
      stages,
      operators: this._operators,
      config_hash
    }
  }
}

File: clients/client-py/queen/streams/stream.py
"""
Stream — fluent builder for streaming pipelines. 1:1 port of Stream.js.

A Stream is an immutable chain of Operators that gets compiled and run by a
Runner. Each combinator (.map, .filter, .gate, ...) returns a NEW Stream
with the operator appended; the chain is finalized by ``await stream.run(...)``.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Callable, List, Optional

from .operators.aggregate_op import AggregateOperator
from .operators.filter_op import FilterOperator
from .operators.flat_map_op import FlatMapOperator
from .operators.foreach_op import ForeachOperator
from .operators.gate_op import GateOperator
from .operators.key_by_op import KeyByOperator
from .operators.map_op import MapOperator
from .operators.reduce_op import ReduceOperator
from .operators.sink_op import SinkOperator
from .operators.window_cron import WindowCronOperator
from .operators.window_session import WindowSessionOperator
from .operators.window_sliding import WindowSlidingOperator
from .operators.window_tumbling import WindowTumblingOperator
from .runtime.runner import Runner
from .util.config_hash import config_hash_of


@dataclass
class _CompiledStages:
    pre: list
    key_by: Optional[Any]
    window: Optional[Any]
    reducer: Optional[Any]
    gate: Optional[Any]
    post: list
    sink: Optional[Any]


@dataclass
class _CompiledStream:
    source: Any
    source_options: dict
    stages: _CompiledStages
    operators: list
    config_hash: str


class Stream:
    """Fluent, immutable streaming pipeline builder."""

    @classmethod
    def from_(cls, queue_builder: Any, **options: Any) -> "Stream":
        """
        Build a Stream sourced from a Queen QueueBuilder.

        Args:
            queue_builder: result of ``queen.queue('name')`` — must
                expose ``.pop()``, ``.ack()``, ``.push()`` and the queue name.
        """
        return cls(source=queue_builder, operators=[], source_options=options)

    # JS-compatible alias
    @classmethod
    def from_queue(cls, queue_builder: Any, **options: Any) -> "Stream":
        return cls.from_(queue_builder, **options)

    def __init__(self, source: Any, operators: List[Any], source_options: Optional[dict] = None):
        self._source = source
        self._source_options = source_options or {}
        self._operators = operators

    # ---------------------------------------------------------------- Stateless

    def map(self, fn: Callable[..., Any]) -> "Stream":
        return self._extend(MapOperator(fn))

    def filter(self, predicate: Callable[..., Any]) -> "Stream":
        return self._extend(FilterOperator(predicate))

    def flat_map(self, fn: Callable[..., Any]) -> "Stream":
        return self._extend(FlatMapOperator(fn))

    # JS aliases (camelCase) for portability
    flatMap = flat_map

    # -------------------------------------------------------------------- Keying

    def key_by(self, fn: Callable[[Any], Any]) -> "Stream":
        return self._extend(KeyByOperator(fn))

    keyBy = key_by

    # -------------------------------------------------------------------- Windows

    def window_tumbling(self, **opts: Any) -> "Stream":
        return self._extend(WindowTumblingOperator(**_kwargs_camel_to_snake(opts)))

    def window_sliding(self, **opts: Any) -> "Stream":
        return self._extend(WindowSlidingOperator(**_kwargs_camel_to_snake(opts)))

    def window_session(self, **opts: Any) -> "Stream":
        return self._extend(WindowSessionOperator(**_kwargs_camel_to_snake(opts)))

    def window_cron(self, **opts: Any) -> "Stream":
        return self._extend(WindowCronOperator(**_kwargs_camel_to_snake(opts)))

    # JS aliases
    windowTumbling = window_tumbling
    windowSliding = window_sliding
    windowSession = window_session
    windowCron = window_cron

    # -------------------------------------------------------------------- Reduce

    def reduce(self, fn: Callable, initial: Any = None) -> "Stream":
        return self._extend(ReduceOperator(fn, initial))

    def aggregate(self, extractors: dict) -> "Stream":
        return self._extend(AggregateOperator(extractors))

    # ---------------------------------------------------------------------- Gate

    def gate(self, fn: Callable[..., Any]) -> "Stream":
        """
        Per-message ALLOW/DENY decision with persistent per-key state.

        Used to build rate limiters, throttlers, fairness gates, circuit
        breakers. See queen.streams.helpers.rate_limiter for ready-made
        gate factories.
        """
        return self._extend(GateOperator(fn))

    # ---------------------------------------------------------------------- Sink

    def to(self, sink_queue_builder: Any, partition: Optional[Any] = None) -> "Stream":
        return self._extend(SinkOperator(sink_queue_builder, partition=partition))

    def foreach(self, fn: Callable[..., Any]) -> "Stream":
        return self._extend(ForeachOperator(fn))

    # ----------------------------------------------------------------- Compile/run

    async def run(self, query_id: str, url: str, **opts: Any) -> Runner:
        """
        Start the streaming runner. Returns a Runner exposing ``stop()`` and
        ``metrics()``.

        Args:
            query_id: durable identity of this query (becomes the
                consumer group + state shard).
            url: Queen broker URL (e.g. ``http://localhost:6632``).
            batch_size: messages per cycle (default 200)
            max_partitions: pop up to N partitions per cycle (default 4)
            max_wait_millis: long-poll wait for source pop (default 1000)
            subscription_mode: 'all' (default) | 'new'
            subscription_from: ISO timestamp or 'now'
            reset: True to wipe existing state if config_hash mismatches
            consumer_group: override the default 'streams.<queryId>' group
            on_error: callable(err, ctx) for cycle error hook
            bearer_token: JWT for auth-enabled brokers
        """
        if not query_id:
            raise ValueError("run requires query_id=...")
        compiled = self._compile()
        runner_opts: dict = {
            "queryId": query_id,
            "url": url,
            "stream": compiled,
        }
        # Map snake_case kwargs to the runner's camelCase keys.
        for k, v in opts.items():
            runner_opts[_to_camel(k)] = v
        runner = Runner(**runner_opts)
        await runner.start()
        return runner

    # ===== internals =========================================================

    def _extend(self, operator: Any) -> "Stream":
        return Stream(
            source=self._source,
            operators=[*self._operators, operator],
            source_options=self._source_options,
        )

    def _compile(self) -> _CompiledStream:
        sink = next((op for op in self._operators if op.kind in ("sink", "foreach")), None)
        sink_idx = self._operators.index(sink) if sink else len(self._operators)
        if sink and sink_idx != len(self._operators) - 1:
            raise ValueError(
                "sink operators (.to / .foreach) must be the last in the chain; "
                f"found {sink.kind} at position {sink_idx} of {len(self._operators)}"
            )
        upstream = self._operators[:sink_idx] if sink else list(self._operators)

        stages = _CompiledStages(
            pre=[], key_by=None, window=None, reducer=None, gate=None, post=[], sink=sink
        )
        phase = "pre"
        for op in upstream:
            kind = op.kind
            if kind in ("map", "filter", "flatMap"):
                if phase in ("pre", "keyed", "window"):
                    stages.pre.append(op)
                else:
                    stages.post.append(op)
            elif kind == "keyBy":
                if stages.key_by is not None:
                    raise ValueError("only one .key_by() per stream")
                if phase in ("reducer", "gate"):
                    raise ValueError(".key_by() must come before window/reduce/gate")
                stages.key_by = op
                if phase == "pre":
                    phase = "keyed"
            elif kind == "window":
                if stages.window is not None:
                    raise ValueError("only one window operator per stream")
                if phase == "reducer":
                    raise ValueError("window must come before reduce/aggregate")
                if stages.gate is not None:
                    raise ValueError("window+reduce is incompatible with .gate() in the same stream")
                stages.window = op
                phase = "window"
            elif kind in ("reduce", "aggregate"):
                if stages.reducer is not None:
                    raise ValueError("only one reduce/aggregate per stream")
                if stages.window is None:
                    raise ValueError("reduce/aggregate requires a preceding window operator")
                if stages.gate is not None:
                    raise ValueError("reduce/aggregate is incompatible with .gate() in the same stream")
                stages.reducer = op
                phase = "reducer"
            elif kind == "gate":
                if stages.gate is not None:
                    raise ValueError("only one .gate() per stream")
                if stages.window is not None or stages.reducer is not None:
                    raise ValueError(".gate() is incompatible with windowing/reduce in the same stream")
                stages.gate = op
                phase = "gate"
            else:
                raise ValueError(f"unsupported operator: {kind}")

        config_hash = config_hash_of(self._operators)

        return _CompiledStream(
            source=self._source,
            source_options=self._source_options,
            stages=stages,
            operators=list(self._operators),
            config_hash=config_hash,
        )


def _to_camel(s: str) -> str:
    parts = s.split("_")
    return parts[0] + "".join(p.title() for p in parts[1:])


def _kwargs_camel_to_snake(opts: dict) -> dict:
    """
    Accept both camelCase JS-style kwargs and snake_case Python-style for
    window operators. Internally always translate to snake_case so the
    operator constructors can use Python conventions.
    """
    out: dict = {}
    for k, v in opts.items():
        if k == "gracePeriod":
            out["grace_period"] = v
        elif k == "idleFlushMs":
            out["idle_flush_ms"] = v
        elif k == "eventTime":
            out["event_time"] = v
        elif k == "allowedLateness":
            out["allowed_lateness"] = v
        elif k == "onLate":
            out["on_late"] = v
        else:
            out[k] = v
    return out
